### PR TITLE
Improve documentation for `fill`

### PR DESCRIPTION
**Pull request by mbauman**
Ref #41209

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -431,14 +431,74 @@ to_dim(d::Integer) = d
...
-Create an array filled with the value `x`. For example, `fill(1.0, (5,5))` returns a 5×5
-array of floats, with each element initialized to `1.0`.
+Create an array of size `dims` with every element set to `value`.
 
-`dims` may be specified as either a tuple or a sequence of arguments. For example,
-the common idiom `fill(x)` creates a zero-dimensional array containing the single value `x`.
+For example, `fill(1.0, (5,5))` returns a 5×5 array of floats,
+with `1.0` in every location of the array.
+
+The dimension lengths `dims` may be specified as either a tuple or a sequence of arguments.
```
Comment (mbauman) [inline base/array.jl:442]:
```suggestion
The dimension lengths `dims` may be specified as either a tuple or a sequence of arguments.
```

**Comment by CameronBieganek** [comment]
Much better! Thanks!

I think the array comprehension example makes it a lot more clear what a person needs to do if they don't want an array filled with references to the same object.

**Review by simeonschaub — APPROVED** [review]
Nice, I think this is much clearer! Thanks for writing this up!

**Comment by CameronBieganek** [comment]
I like how this avoids using the word "reference". As far as I'm aware, "reference" (with a lower case "r") is not a formally defined concept in Julia. The closest thing is a variable binding, which binds a *variable* to a *value*.

File: base/array.jl [hunk 2]
```diff
@@ -431,14 +431,74 @@ to_dim(d::Integer) = d
...
+An `N`-length tuple or `N` arguments following the `value` specify an `N`-dimensional
+array. Thus, a common idiom for creating a zero-dimensional array with its only value
```
Comment (mbauman) [inline base/array.jl:444]:
```suggestion
array. Thus, a common idiom for creating a zero-dimensional array with its only location
```

File: base/array.jl [hunk 3]
```diff
@@ -431,14 +431,74 @@ to_dim(d::Integer) = d
...
 
 """
-    fill(x, dims::Tuple)
-    fill(x, dims...)
+    fill(value, dims::Tuple)
+    fill(value, dims...)
 
-Create an array filled with the value `x`. For example, `fill(1.0, (5,5))` returns a 5×5
-array of floats, with each element initialized to `1.0`.
+Create an array of size `dims` with every element set to `value`.
```
Comment (mbauman) [inline base/array.jl:437]:
```suggestion
Create an array of size `dims` with every location set to `value`.
```

File: base/array.jl [hunk 4]
```diff
@@ -431,14 +431,74 @@ to_dim(d::Integer) = d
...
-the common idiom `fill(x)` creates a zero-dimensional array containing the single value `x`.
+For example, `fill(1.0, (5,5))` returns a 5×5 array of floats,
+with `1.0` in every location of the array.
+
+The dimension lengths `dims` may be specified as either a tuple or a sequence of arguments.
+An `N`-length tuple or `N` arguments following the `value` specify an `N`-dimensional
+array. Thus, a common idiom for creating a zero-dimensional array with its only value
+set to `x` is `fill(x)`.
+
+Every element of the returned array is set to (and is thus [`===`](@ref) to)
```
Comment (mbauman) [inline base/array.jl:447]:
```suggestion
Every location of the returned array is set to (and is thus [`===`](@ref) to)
```

**Review by mbauman — COMMENTED** [review]
Be more consistent about setting _locations_ in the array (to not be confused with modifying an element in a given location)

File: base/array.jl [hunk 5]
```diff
@@ -431,14 +431,74 @@ to_dim(d::Integer) = d
...
+ 8675309
+
+julia> v
+3-element Vector{Vector{Any}}:
+ [8675309]
+ [8675309]
+ [8675309]
+```
+
+To create an array of many independent inner arrays, use a [comprehension](@ref man-comprehension) instead.
```
Comment (mbauman) [inline base/array.jl:479]:
```suggestion
To create an array of many independent inner arrays, use a [comprehension](@ref man-comprehensions) instead.
```

**Review by ararslan — APPROVED** [review]
image

**Pull request merged**